The [failure in ci-runtime](https://github.com/cilium/cilium/actions/runs/9361183964/job/25782047708#step:17:362) is in a related area:

```
--- FAIL: TestDevicesController_DeviceDetection (0.01s)
361
    logger.go:257: time=2024-06-04T11:12:55.923Z level=INFO source=/host/vendor/github.com/cilium/hive/hive.go:337 msg=Starting
362
    logger.go:257: time=2024-06-04T11:12:55.923Z level=INFO source=/host/pkg/datapath/linux/devices_controller.go:522 msg="Devices changed" module=devices-controller devices="[dummy0 dummy1 eth0]"
363
    logger.go:257: time=2024-06-04T11:12:55.924Z level=INFO source=/host/vendor/github.com/cilium/hive/hive.go:341 msg=Started duration=785.306µs
364
    logger.go:257: time=2024-06-04T11:12:55.924Z level=INFO source=/host/vendor/github.com/cilium/hive/hive.go:353 msg=Stopping
365
    logger.go:257: time=2024-06-04T11:12:55.924Z level=INFO source=/host/pkg/hive/health/provider.go:62 msg="health.job-module-status-metrics (rev=2)" module=health
366
    logger.go:257: time=2024-06-04T11:12:55.924Z level=INFO source=/host/vendor/github.com/cilium/hive/hive.go:337 msg=Starting
367
    logger.go:257: time=2024-06-04T11:12:55.924Z level=INFO source=/host/vendor/github.com/cilium/hive/hive.go:341 msg=Started duration=16.391µs
368
    logger.go:257: time=2024-06-04T11:12:55.925Z level=INFO source=/host/pkg/datapath/linux/devices_controller.go:522 msg="Devices changed" module=devices-controller devices="[eth0 dummy0 dummy1]"
369
    devices_controller_test.go:473: 
370
        	Error Trace:	/host/pkg/datapath/linux/devices_controller_test.go:473
371
        	            				/host/pkg/netns/netns_linux.go:171
372
        	            				/host/vendor/golang.org/x/sync/errgroup/errgroup.go:78
373
        	            				/root/sdk/go1.22.3/src/runtime/asm_amd64.s:1695
374
        	Error:      	Should be true
375
        	Test:       	TestDevicesController_DeviceDetection
```

If the failure is introduced by this PR, then it will reoccur if we re-run the tests. I can try that.

--- a/Documentation/cmdref/cilium-agent.md
+++ b/Documentation/cmdref/cilium-agent.md
@@ -185,6 +185,7 @@ cilium-agent [flags]
       --exclude-node-label-patterns strings                       List of k8s node label regex patterns to be excluded from CiliumNode
       --external-envoy-proxy                                      whether the Envoy is deployed externally in form of a DaemonSet or not
       --fixed-identity-mapping map                                Key-value for the fixed identity mapping which allows to use reserved label for fixed identities, e.g. 128=kv-store,129=kube-dns
+      --force-device-detection                                    Forces the auto-detection of devices, even if specific devices are explicitly listed
       --gateway-api-secrets-namespace string                      GatewayAPISecretsNamespace is the namespace having tls secrets used by CEC, originating from Gateway API
       --gops-port uint16                                          Port for gops server to listen on (default 9890)
   -h, --help                                                      help for cilium-agent

--- a/Documentation/cmdref/cilium-agent_hive.md
+++ b/Documentation/cmdref/cilium-agent_hive.md
@@ -51,6 +51,7 @@ cilium-agent hive [flags]
       --envoy-keep-cap-netbindservice                             Keep capability NET_BIND_SERVICE for Envoy process
       --envoy-log string                                          Path to a separate Envoy log file, if any
       --envoy-secrets-namespace string                            EnvoySecretsNamespace is the namespace having secrets used by CEC
+      --force-device-detection                                    Forces the auto-detection of devices, even if specific devices are explicitly listed
       --gateway-api-secrets-namespace string                      GatewayAPISecretsNamespace is the namespace having tls secrets used by CEC, originating from Gateway API
       --gops-port uint16                                          Port for gops server to listen on (default 9890)
   -h, --help                                                      help for hive

--- a/Documentation/cmdref/cilium-agent_hive_dot-graph.md
+++ b/Documentation/cmdref/cilium-agent_hive_dot-graph.md
@@ -57,6 +57,7 @@ cilium-agent hive dot-graph [flags]
       --envoy-keep-cap-netbindservice                             Keep capability NET_BIND_SERVICE for Envoy process
       --envoy-log string                                          Path to a separate Envoy log file, if any
       --envoy-secrets-namespace string                            EnvoySecretsNamespace is the namespace having secrets used by CEC
+      --force-device-detection                                    Forces the auto-detection of devices, even if specific devices are explicitly listed
       --gateway-api-secrets-namespace string                      GatewayAPISecretsNamespace is the namespace having tls secrets used by CEC, originating from Gateway API
       --gops-port uint16                                          Port for gops server to listen on (default 9890)
       --http-idle-timeout uint                                    Time after which a non-gRPC HTTP stream is considered failed unless traffic in the stream has been processed (in seconds); defaults to 0 (unlimited)

--- a/Documentation/helm-values.rst
+++ b/Documentation/helm-values.rst
@@ -1492,6 +1492,10 @@
      - Additional agent volumes.
      - list
      - ``[]``
+   * - :spelling:ignore:`forceDeviceDetection`
+     - Forces the auto-detection of devices, even if specific devices are explicitly listed
+     - bool
+     - ``false``
    * - :spelling:ignore:`gatewayAPI.enableAlpn`
      - Enable ALPN for all listeners configured with Gateway API. ALPN will attempt HTTP/2, then HTTP 1.1. Note that this will also enable ``appProtocol`` support, and services that wish to use HTTP/2 will need to indicate that via their ``appProtocol``.
      - bool

--- a/install/kubernetes/cilium/README.md
+++ b/install/kubernetes/cilium/README.md
@@ -423,6 +423,7 @@ contributors across the globe, there is almost always someone available to help.
 | extraInitContainers | list | `[]` | Additional initContainers added to the cilium Daemonset. |
 | extraVolumeMounts | list | `[]` | Additional agent volumeMounts. |
 | extraVolumes | list | `[]` | Additional agent volumes. |
+| forceDeviceDetection | bool | `false` | Forces the auto-detection of devices, even if specific devices are explicitly listed |
 | gatewayAPI.enableAlpn | bool | `false` | Enable ALPN for all listeners configured with Gateway API. ALPN will attempt HTTP/2, then HTTP 1.1. Note that this will also enable `appProtocol` support, and services that wish to use HTTP/2 will need to indicate that via their `appProtocol`. |
 | gatewayAPI.enableAppProtocol | bool | `false` | Enable Backend Protocol selection support (GEP-1911) for Gateway API via appProtocol. |
 | gatewayAPI.enableProxyProtocol | bool | `false` | Enable proxy protocol for all GatewayAPI listeners. Note that _only_ Proxy protocol traffic will be accepted once this is enabled. |

--- a/install/kubernetes/cilium/templates/cilium-configmap.yaml
+++ b/install/kubernetes/cilium/templates/cilium-configmap.yaml
@@ -701,6 +701,10 @@ data:
   enable-runtime-device-detection: "true"
 {{- end }}
 
+{{- if .Values.forceDeviceDetection }}
+  force-device-detection: "true"
+{{- end }}
+
   kube-proxy-replacement: {{ $kubeProxyReplacement | quote }}
 
 {{- if ne $kubeProxyReplacement "disabled" }}

--- a/install/kubernetes/cilium/values.schema.json
+++ b/install/kubernetes/cilium/values.schema.json
@@ -2375,6 +2375,9 @@
       "items": {},
       "type": "array"
     },
+    "forceDeviceDetection": {
+      "type": "boolean"
+    },
     "gatewayAPI": {
       "properties": {
         "enableAlpn": {

--- a/install/kubernetes/cilium/values.yaml
+++ b/install/kubernetes/cilium/values.yaml
@@ -688,6 +688,8 @@ daemon:
 #
 # This option has been deprecated and is a no-op.
 enableRuntimeDeviceDetection: true
+# -- Forces the auto-detection of devices, even if specific devices are explicitly listed
+forceDeviceDetection: false
 # -- Chains to ignore when installing feeder rules.
 # disableIptablesFeederRules: ""
 

--- a/install/kubernetes/cilium/values.yaml.tmpl
+++ b/install/kubernetes/cilium/values.yaml.tmpl
@@ -687,6 +687,9 @@ daemon:
 # This option has been deprecated and is a no-op.
 enableRuntimeDeviceDetection: true
 
+# -- Forces the auto-detection of devices, even if specific devices are explicitly listed
+forceDeviceDetection: false
+
 # -- Chains to ignore when installing feeder rules.
 # disableIptablesFeederRules: ""
 

--- a/pkg/datapath/linux/devices_controller.go
+++ b/pkg/datapath/linux/devices_controller.go
@@ -71,6 +71,8 @@ var DevicesControllerCell = cell.Module(
 
 func (c DevicesConfig) Flags(flags *pflag.FlagSet) {
 	flags.StringSlice(option.Devices, []string{}, "List of devices facing cluster/external network (used for BPF NodePort, BPF masquerading and host firewall); supports '+' as wildcard in device name, e.g. 'eth+'")
+
+	flags.Bool(option.ForceDeviceDetection, false, "Forces the auto-detection of devices, even if specific devices are explicitly listed")
 }
 
 var (
@@ -95,6 +97,9 @@ type DevicesConfig struct {
 	// If empty the devices are auto-detected according to rules defined
 	// by isSelectedDevice().
 	Devices []string
+	// ForceDeviceDetection forces the auto-detection of devices,
+	// even if user-specific devices are explicitly listed.
+	ForceDeviceDetection bool
 }
 
 type devicesControllerParams struct {
@@ -114,9 +119,10 @@ type devicesController struct {
 	params devicesControllerParams
 	log    *slog.Logger
 
-	initialized    chan struct{}
-	filter         deviceFilter
-	l3DevSupported bool
+	initialized          chan struct{}
+	filter               deviceFilter
+	enforceAutoDetection bool
+	l3DevSupported       bool
 
 	// deadLinkIndexes tracks the set of links that have been deleted. This is needed
 	// to avoid processing route or address updates after a link delete as they may
@@ -128,11 +134,12 @@ type devicesController struct {
 
 func newDevicesController(lc cell.Lifecycle, p devicesControllerParams) (*devicesController, statedb.Table[*tables.Device], statedb.Table[*tables.Route]) {
 	dc := &devicesController{
-		params:          p,
-		initialized:     make(chan struct{}),
-		filter:          deviceFilter(p.Config.Devices),
-		log:             p.Log,
-		deadLinkIndexes: sets.New[int](),
+		params:               p,
+		initialized:          make(chan struct{}),
+		filter:               deviceFilter(p.Config.Devices),
+		enforceAutoDetection: p.Config.ForceDeviceDetection,
+		log:                  p.Log,
+		deadLinkIndexes:      sets.New[int](),
 	}
 	lc.Append(dc)
 	return dc, p.DeviceTable, p.RouteTable
@@ -543,12 +550,15 @@ func (dc *devicesController) isSelectedDevice(d *tables.Device, txn statedb.Writ
 	}
 
 	// If user specified devices or wildcards, then skip the device if it doesn't match.
-	// If the device does match, then skip further checks.
+	// If the device does match and user not requested auto detection, then skip further checks.
+	// If the device does match and user requested auto detection, then continue to further checks.
 	if dc.filter.nonEmpty() {
 		if dc.filter.match(d.Name) {
 			return true, ""
 		}
-		return false, fmt.Sprintf("not matching user filter %v", dc.filter)
+		if !dc.enforceAutoDetection {
+			return false, fmt.Sprintf("not matching user filter %v", dc.filter)
+		}
 	}
 
 	// Skip devices that have an excluded interface flag set.

--- a/pkg/datapath/linux/devices_controller_test.go
+++ b/pkg/datapath/linux/devices_controller_test.go
@@ -386,6 +386,10 @@ func TestDevicesController_Wildcards(t *testing.T) {
 		require.NoError(t, createDummy("dummy0", "192.168.0.1/24", false))
 		require.NoError(t, createDummy("nonviable", "192.168.1.1/24", false))
 
+		// This device satisfies the autodetection rule, but should not be included
+		// because the ForceDeviceDetection option is not enabled
+		require.NoError(t, createDummy("eth0", "1.2.3.4/24", false))
+
 		for {
 			rxn := db.ReadTxn()
 			devs, invalidated := tables.SelectedDevices(devicesTable, rxn)
@@ -404,6 +408,76 @@ func TestDevicesController_Wildcards(t *testing.T) {
 
 		err = h.Stop(tlog, context.TODO())
 		assert.NoError(t, err)
+		return nil
+	})
+}
+
+// TestDevicesController_with_ForcedDetection tests the behavior of device detection when forced detection is enabled.
+// It expects all devices matching a specific pattern to be detected will append to detected devices and marked as selected.
+func TestDevicesController_with_ForcedDetection(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	testutils.PrivilegedTest(t)
+	devicesControllerTestSetup(t)
+
+	tlog := hivetest.Logger(t)
+	ns := netns.NewNetNS(t)
+	ns.Do(func() error {
+		var (
+			db           *statedb.DB
+			devicesTable statedb.Table[*tables.Device]
+			h            *hive.Hive
+		)
+
+		// Function to set up the hive and run device detection
+		runDeviceDetection := func(devicePattern string, forceDetection bool) error {
+			h = hive.New(
+				DevicesControllerCell,
+				cell.Provide(func() (*netlinkFuncs, error) { return makeNetlinkFuncs() }),
+				cell.Invoke(func(db_ *statedb.DB, devicesTable_ statedb.Table[*tables.Device]) {
+					db = db_
+					devicesTable = devicesTable_
+				}),
+			)
+			hive.AddConfigOverride(h, func(c *DevicesConfig) {
+				c.Devices = []string{devicePattern}
+				c.ForceDeviceDetection = forceDetection
+			})
+
+			return h.Start(tlog, ctx)
+		}
+
+		// Function to check the expected number of devices
+		testDevices := func(expectedCount int) bool {
+			rxn := db.ReadTxn()
+			devs, invalidated := tables.SelectedDevices(devicesTable, rxn)
+			if len(devs) == expectedCount {
+				return true
+			}
+
+			select {
+			case <-ctx.Done():
+				t.Fatalf("Test timed out while waiting for devices, last seen: %v", devs)
+				return false
+			case <-invalidated:
+				return false
+			}
+		}
+
+		// Create dummy interfaces as per test requirements
+		require.NoError(t, createDummy("dummy0", "192.168.0.1/24", false))
+		require.NoError(t, createDummy("dummy1", "192.168.1.1/24", false))
+
+		// This device does not match the "dummy+" pattern, but should be included
+		// because the ForceDeviceDetection option is enabled
+		require.NoError(t, createDummy("eth0", "1.2.3.4/24", false))
+
+		// Test with forced detection enabled
+		require.NoError(t, runDeviceDetection("dummy+", true))
+		require.True(t, testDevices(3), "Expecting all three devices to be detected")
+		require.NoError(t, h.Stop(tlog, ctx))
+
 		return nil
 	})
 }

--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -123,6 +123,9 @@ const (
 	// Devices facing cluster/external network for attaching bpf_host
 	Devices = "devices"
 
+	// Forces the auto-detection of devices, even if specific devices are explicitly listed
+	ForceDeviceDetection = "force-device-detection"
+
 	// DirectRoutingDevice is the name of a device used to connect nodes in
 	// direct routing mode (only required by BPF NodePort)
 	DirectRoutingDevice = "direct-routing-device"

